account/src/main/java/kr/flab/movieon/account/infrastructure/AuthenticateAccountProcessorImpl.java

```diff
@@ -0,0 +1,29 @@
+package kr.flab.movieon.account.infrastructure;
+
+import kr.flab.movieon.account.domain.AuthenticateAccountProcessor;
+import kr.flab.movieon.account.infrastructure.security.domain.AccountContext;
+import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+public final class AuthenticateAccountProcessorImpl implements AuthenticateAccountProcessor {
+
```

Comment:
username와 password의 일치 여부는 검사하지 않는 건가요?

Reply:
단순히 패스워드를 비교하는 로직이라 따로 Provider를 만들지 않고, 시큐리티의 기본 프로바이더로 처리했습니다.
나중에 추가적인 인증이 필요하게 되면 
`public class CustomAuthenticationProvider implements AuthenticationProvider` 같은 걸 추가하도록 하겠습니다 😀

-------------

`시큐리티가 알아서 해준다`라고 설명하니 조금 불안해서, 직접 코드를 뜯어봤는데

### 1. 로그인 과정에서 아래 코드를 실행합니다.
 - `authenticationManager`의 구현체인 `ProviderManager`를 통해 해당 토큰을 처리할 수 있는 프로바이더를 찾습니다.
```java
Authentication authentication = authenticationManager.authenticate(
            new UsernamePasswordAuthenticationToken(username, password)
        );
```
 
</br>

### 2. 따로 등록해둔 프로바이더가 없기 때문에, 기본 프로바이더인 `DaoAuthenticationProvider` 가 동작합니다.
- 이 프로바이더는 매니저에 등록된 `UserDetailsService`와 `PasswordEncoder`를 사용합니다.
```java
    // SecurityConfig.java
    @Override
    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
        auth.userDetailsService(accountDetailsService).passwordEncoder(passwordEncoder());
        auth.authenticationProvider(jwtAuthenticationProvider); // 이건 JwtAuthToken 을 처리하는 Provider
    }
```
</br>

- DaoAuthenticationProvider 는 아래 클래스를 상속하며,  `UsernamePasswordAuthenticationToken`  를 support 합니다.

```java
public abstract class AbstractUserDetailsAuthenticationProvider{
... 생략 ...
	@Override
	public boolean supports(Class<?> authentication) {
		return (UsernamePasswordAuthenticationToken.class.isAssignableFrom(authentication));
	}

	@Override
	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
        ...
      try {
		      this.preAuthenticationChecks.check(user);
	              additionalAuthenticationChecks(user, (UsernamePasswordAuthenticationToken) authentication);
	 	  } 
	    catch (AuthenticationException ex) {
            ...
          }
        return createSuccessAuthentication(principalToReturn, authentication, user);
      }
}
```

</br>

### 3. 즉, `Provider.authenticate(...)`가 실행되면, 아래의 로직이 실행됩니다.

```java
public class DaoAuthenticationProvider extends AbstractUserDetailsAuthenticationProvider {
... 생략 ...
	protected void additionalAuthenticationChecks(UserDetails userDetails,
			UsernamePasswordAuthenticationToken authentication) throws AuthenticationException {
		if (authentication.getCredentials() == null) {
			this.logger.debug("Failed to authenticate since no credentials provided");
			throw new BadCredentialsException(this.messages
					.getMessage("AbstractUserDetailsAuthenticationProvider.badCredentials", "Bad credentials"));
		}
		String presentedPassword = authentication.getCredentials().toString();
		if (!this.passwordEncoder.matches(presentedPassword, userDetails.getPassword())) {
			this.logger.debug("Failed to authenticate since password does not match stored value");
			throw new BadCredentialsException(this.messages
					.getMessage("AbstractUserDetailsAuthenticationProvider.badCredentials", "Bad credentials"));
		}
	}
...
}	

```